### DISCUSSION TITLE
Growatt DTC collection - Help wanted

### DISCUSSION THREAD

**Discussion by egguy** [johanmeijer/grott#201] [Q&A]
Hello,

I'm working on implementing some new functionality on groot (like direct HA integration). At the moment, the discovery of the layout on the grott side is partially manual for SPH (need to configure invtype=sph).

I think it should be possible to discover the model and by association the decoding layout by analysing the DTC field of the 0×0103 packet.

I found some documentation on the Modbus RTU documentation (and other obscure docs).

| Value | Type       | Note                         |
|-------|------------|------------------------------|
| 2049 | Inverter | 10000TL3-S |
| 031xx | PV Storage | Front 1 - tracker PV Storage |
| 034XX | OffGrid    | OffGrid SPF 3 ‐ 5K           |
| 3501  | OffGrid    | SPH 6000                     |
| 5000  | inverter | TL3-MAX                         |
| 5001  | inverter | TL3-MID                       |

I only have one inverter to work with (the SPH 6000).

I would like to collect multiple version to help me create an auto-detection code for the layout.

## How to contribute 

The DTC value is located in the register 43, this value can be accessed directly online from the growatt interface.

This is how you can do it:
- Go to your plant interface -> At the bottom there is "My photovoltaic devices' -> click on settings (accept the disclaimer)
- Go to the bottom -> click on advanced Set -> Select "Start address" -> put 43 and 43 in the 2 fields
- The Key is the current date in the format: YYYYMMDD, for the 31 october 2022 -> 20221031
- Then click on Advanced reader, after a while the "Current Value will appear" 

![image](https://user-images.githubusercontent.com/186777/199014972-9ef1e175-7d5a-4ccf-a3a6-eb4b159787b4.png)

After the value has been obtained, could you share it with me in for the format:

value - Device model 

Ex: 3501 - SPH6000

Thx a lot.

**Comment by johanmeijer** [comment]
Yes full automated detection is the ultimate goal! 

But there are what challenges with that at this moment. 

During startup Grott needs to identify the inverter type (probably also the datalogger and the combination of those) and the store it somewhere  (at this moment grott is stateless and does not store status). 

I would be nice if the device type is in the data record but I have not been able to detect that yet. With grottserver an API is available to query the registers (the same as you use via the the growatt website). By there is no link between grott and grottserver yet (will prably come in Version 3). 

I will try to find the inverter types in the data or announce records and see if I can use that within grott. Do not expect a quick fix while it needs a lot of design an build. But it absolutely a good improvement if Grott will be able to do this. 

To the others users please send me your inverter and datalogger types and the value of reg 43 (although it does not have to be the same register for al inverters). But the more information we have the easier it will be to decode.

**Comment by cyberdiamond** [comment]
Tested 1x 10000TL3-S

Current Value(43~43)：2049


Also tested 2x SPH3~6k TL BL (both are SPH5000's with different
firmware)

Current Value(43~43)：3501

So this is the same as your SPH6000 as expected being in the same
family but good to have the data confirmed.

Cheers,
Steve

-----Original Message-----
From: Johan Meijer ***@***.***>
Reply-To: johanmeijer/grott
***@***.***>
To: johanmeijer/grott ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [johanmeijer/grott] Growatt DTC collection - Help wanted
(Discussion #201)
Date: 01/11/22 06:42:50

Yes full automated detection is the ultimate goal!
But there are what challenges with that at this moment.
During startup Grott needs to identify the inverter type (probably also
the datalogger and the combination of those) and the store it somewhere
(at this moment grott is stateless and does not store status).
I would be nice if the device type is in the data record but I have not
been able to detect that yet. With grottserver an API is available to
query the registers (the same as you use via the the growatt website).
By there is no link between grott and grottserver yet (will prably come
in Version 3).
I will try to find the inverter types in the data or announce records
and see if I can use that within grott. Do not expect a quick fix while
it needs a lot of design an build. But it absolutely a good improvement
if Grott will be able to do this.
To the others users please send me your inverter and datalogger types
and the value of reg 43 (although it does not have to be the same
register for al inverters). But the more information we have the easier
it will de to decode.
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you are subscribed to this
thread.Message ID: <johanmeijer/grott/repo-
***@***.***>

**Reply by egguy** [reply]
Thank you, I've updated the table.

**Reply by jupppo** [reply]
My SHP3600 has also: {"value": 3501}

**Comment by githubDante** [comment]
Hi,

These values does not align with the documentation at all, which states the value must be in format 001xx, 002xx etc. When read as integer indeed the value is 3501 for SPH. Below are the values from MID40KTL3-X & MAX 80KTL3 LV (not related to this plugin, but probably good addition to the table at the top).
```
>>> check_543.inverter
<InverterType.MID: 'mid'>
>>> check_543.report
True
>>> check_543.int_at(43)
5001
>>> check_643.inverter
<InverterType.MAX: 'max'>
>>> check_643.report
True
>>> check_643.int_at(43)
5000
```

**Reply by egguy** [reply]
There's a PDF floating around with 2 more values for DTC, it seems to match:
![image](https://user-images.githubusercontent.com/186777/202280943-fda6a498-609f-4943-a77a-bf84ec5ff270.png)

It seems the DTC is a range, like:
- 3500 > DTC > 3400 == SPH 
- 3400 > DTC > 3300 == SPF

If I'm not mistaken, it can be found in DATA03 (0x0103) packet

This is why my hunch is that the DTC could be a good indicator of the type of inverter. On the other hand, as @johanmeijer said it, if you can find it in DATA03 packet, you will need to make the grottproxy a tad more stateful to track the model by TCP session.

**Reply by githubDante** [reply]
Found them in the SPF documentation. OK, TL3 is somewhere in the 5000 range 😃 
Edit: at least the MID & MAX series, there are many models for which the suffix is TL3

**Reply by egguy** [reply]
Ok, I've updated the table.

**Reply by githubDante** [reply]
> This is why my hunch is that the DTC could be a good indicator of the type of inverter. On the other hand, as @johanmeijer said it, if you can find it in DATA03 packet, you will need to make the grottproxy a tad more stateful to track the model by TCP session.

It's not stateless even at the moment (despite that I use my Proxy Poll version). The connection between the datalogger and the server is persistent. There is a lot of chat on reconnect (interesting packets indeed 😃 ). Already working on an async version of the proxy where each datalogger connection has its own state (session). I do not know if it will be accepted here, but I need it anyway.

**Reply by egguy** [reply]
I think it was more a reference to how the DATA4 Packet is decoded (and guessed) in the procdata function, than the Proxy part.

I don't know if you have found this link ? https://www.vromans.org/johan/software/sw_growatt_wifi_protocol.html but it's a good reference of the packets.

**Reply by githubDante** [reply]
> I don't know if you have found this link ? https://www.vromans.org/johan/software/sw_growatt_wifi_protocol.html but it's a good reference of the packets.

No, first time I see it. I already have the packets enumerated.
I should disagree with 1.3.8 CONFIGURE section, as it looks like 0013 sets the date (at least for me), but this needs different discussion 😃

**Comment by johanmeijer** [comment]
Gents what is the problem. I used the protocol document you referred for coding Grottserver. I think most of the records types are recognized and used (confirmed) in the grottserver. 

Grottdata only decode type 4 records (and is stateless). Decoding type 03 (and not yet discussed and nowhere documented 29 records)  was until now not needed. And I am still not convinced we need them. They are only sent at session start up (once a day) and does not add any useful information. 

Grottserver is not stateless, it keeps information about the inverter and datalogger. You can retrieve this info with the api commands (as described in the grottserver discussion topic). 

I have done a lot of analysis of the data records sent. the communication between the datalogger and growatt server is very simple. There is no real info sent about the inverter / datalogger type (so no 43 reg info). Probably this not needed while growatt can recognize this based on the serial nummers (they are the owner and have this kind of information). 

The only command that is sent after session setup is a time command (also being done by grottserver). Next to it the data logger sent sometimes register info to growatt server (but this seems to be a little bit unsolicited, growatt is not requesting this info). 

Just for you info. I started with Grott 4 years ago and have done a lot of investigation about the protocol and data records. Grott (and grottserver) might not be perfect but is absolutely doing it's thing and does this rather well seeing the amount of people using it. 

The fact I am already working 4 years on this project also means I will not be working on it every day. I really like your enthusiasm but I also hope that you understand that I am a little bit hesitant with accepting big changes while constancy and backwards compatibility is very important (also a golden rule for me as an IT professional). Next to it with big code changes I loose the overview over the coding and might hamper my design and ideas about the future design. But good idea's are always welcome. 

By the way grottdata can also be used standalone to test/decode  the layout. That is what I use for testing (see grotttest.py in the examples directory). 

I will not publish records from other users on gitthub (privacy reasons). If you want this please sent me a mail at grott@ledidobe.nl.

**Reply by githubDante** [reply]
As I stated in other discussions you've done great job so far. No offense or anything, we are trying to help with what we can.
The discussions and the information shared here are really helpful. The reasons behind the async version of the proxy are completely different.
